[tf.data] Add grappler pass to hoist data-discarding ops

zhuzilin opened the pull request:
This is a PR from JIZHI, the AI platform in Tencent.

This pr adds a tf.data grappler pass which is used to hoist the data-discarding ops like `shard`, `skip` and `take`. In this way there will be less unnecessary calculation or cache.
For example, this pass will turn this code:
```python
def parse_and_preprocessing(x):
  # very slow

ds = tf.data.Dataset.TFRecordDataset("example.tfrecord")
ds = ds.map(parse_and_preprocessing, num_parallel_calls=10)
ds = ds.cache()
ds = ds.skip(100)
ds = ds.take(1000)
ds = ds.repeat()
```
into
```python
# ...

ds = tf.data.Dataset.TFRecordDataset("example.tfrecord")
ds = ds.skip(100)
ds = ds.take(1000)
ds = ds.map(parse_and_preprocessing, num_parallel_calls=10)
ds = ds.cache()
ds = ds.repeat()
```
This transformation will enable the `cache` and avoid some unnecessary background calculation in `map`. Also, this pass will help to make a consistently processing map possible (in other words, let the `MapDataset` continue to process whenever there is a empty thread instead of scheduling a element only after a `GetNext` call.). Both of these problems are partially discussed in #39992 . 

The design of this pass is basically the same as `noop_elimination` or `inject_prefetch` and its inner logic is:
1. Find a node that will discard data. (Right now there are `take`, `skip` and `shard`).
Let's use the above code as an example:
```
TFRecord -> ParallelMap -> Cache -> | Skip | -> Take -> Repeat
```
2. Find the chain of ops that will not change the order of the input and is connected to the node in step 1. (Right now there are `map`, `prefetch` and `cache`.)
```
TFRecord -> { ParallelMap -> Cache } -> | Skip | -> Take -> Repeat
```
3. Move the node from the end of the chain to the start.
```
TFRecord -> | Skip | -> { ParallelMap -> Cache } -> Take -> Repeat
```
4. If there is an update of the graph, repeat 1 - 3. (For the example, `take` will be moved in the second round).

gently ping @jsimsa 

Thank you for your time on reviewing this pr.

zhuzilin committed:
add hoist data-discarding ops pass for tf.data

jsimsa commented on tensorflow/core/grappler/optimizers/data/hoist_data_discarding_ops.h:
```diff
@@ -0,0 +1,52 @@
+  HoistDataDiscardingOps() = default;
+  ~HoistDataDiscardingOps() override = default;
+
+  string name() const override { return "hoist_data_discarding_ops"; };
```
Rename to `hoist_discard`

zhuzilin replied:
renamed.

jsimsa commented on tensorflow/core/grappler/optimizers/data/hoist_data_discarding_ops.cc:
```diff
@@ -0,0 +1,120 @@
+          continue;
+        }
+        NodeDef hoisted_node = node;
+        if (!absl::StartsWith(node.name(), "hoist_data_dsicarding_op/")) {
```
typo in the string, change it to simply "hoist_discard".

zhuzilin replied:
changed.

jsimsa commented on tensorflow/core/grappler/optimizers/data/hoist_data_discarding_ops.cc:
```diff
@@ -0,0 +1,120 @@
+  *output = item.graph;
+  MutableGraphView graph(output);
+  bool updated;
+  do {
+    updated = false;
+    for (NodeDef node : graph.graph()->node()) {
```
Would it make sense to change the order of the loops to make this more efficient? Have an outer loop for all the nodes and in the inner loop, if the node is a discarding op, traverse the parent pointers as long as cardinality preserving ops are encountered.

zhuzilin replied:
@jsimsa 
I think the current code behaves basically the same as your suggestion. The outer `do .. while` loop is used in case there are new potential op that can be hoist after a round of optimization. For example, if we have
```
TFRecord -> ParallelMap -> Cache -> Skip -> Take -> Repeat
```
The first  round it may only hoist `skip` and we need a second round to hoist `take`.

jsimsa commented on tensorflow/core/grappler/optimizers/data/hoist_data_discarding_ops.cc:
```diff
@@ -0,0 +1,120 @@
+        NodeDef* start = &node;
+        NodeDef* start_parent = graph_utils::GetInputNode(*start, graph);
+        while (IsCardinalityPreserving(*start_parent) &&
+               NumOutputs(*start_parent, graph.graph()) == 1) {
```
What does this check guard against?

zhuzilin replied:
@jsimsa 
I'm not sure if the following code will make the map dataset have 2 outputs:
```python
ds = ...
ds = ds.map(lambda x: x+1)
ds1 = ds.repeat()
ds2 = ds.take(10)
```
If there will be no such issue, I'll remove the check.

jsimsa commented on tensorflow/core/grappler/optimizers/data/hoist_data_discarding_ops.cc:
```diff
@@ -0,0 +1,120 @@
+        if (start->name() == node.name()) {
+          continue;
+        }
+        NodeDef hoisted_node = node;
```
Why do we need to create a new node? I would expect that we can simply adjusts the fanins and fanouts of the existing nodes.

jsimsa commented on tensorflow/python/data/experimental/ops/optimization_options.py:
```diff
@@ -126,6 +126,13 @@ class OptimizationOptions(options.OptionsBase):
       "a sampling dataset. If None, defaults to False.")
 
+  hoist_data_discarding_ops = options.create_option(
+      name="hoist_data_discarding_ops",
```
hoist discard

zhuzilin replied:
renamed.

jsimsa commented on tensorflow/python/data/experimental/ops/optimization_options.py:
```diff
@@ -126,6 +126,13 @@ class OptimizationOptions(options.OptionsBase):
+      ty=bool,
+      docstring=
+      "Whether to hoist ops that will discard data (such as skip, take, shard)"
+      "out of map transformations. If None, defaults to False.")
```
"out of map transformations" is misleading (and incomplete)

zhuzilin replied:
modified.

jsimsa commented on tensorflow/core/grappler/optimizers/data/hoist_data_discarding_ops.cc:
```diff
@@ -0,0 +1,120 @@
+    "ShardDataset", "SkipDataset", "TakeDataset",
+};
+
+constexpr std::array<const char*, 6> kCardinalityPreserving = {
```
There are other cardinality preserving transformations in the core API and it would be good if your implementation works for those too: `concatenate`, `enumerate`, `shuffle`, and `zip`.

zhuzilin replied:
As for `concatenate`, I think it's hard to hoist those ops when we don't know the cardinality figure. And for `shuffle`, I think that the hoisting may change the ouput. I can add support for `zip` (`enumerate`) I guess, however, in that case, it will be a condition `IsZip` and some extra implementation because right now the supported ops are all unary. I wonder if we could first make the basic functionality work and then let me create another pr for zip?

jsimsa reviewed (commented):
Thank you. You should also add a Python test that uses the `assert_next` transformation to check that the hoisting happens (see `tensorflow/python/data/experimental/kernel_tests/optimization/noop_elimination_test.py` for an example).

zhuzilin committed:
rename the pass the hoist_discard

zhuzilin commented on tensorflow/core/grappler/optimizers/data/BUILD:
```diff
@@ -252,6 +253,44 @@ cc_library(
 )
 
+cc_library(
+    name = "hoist_data_discarding_ops",
```
Renamed.

zhuzilin commented on tensorflow/core/grappler/optimizers/data/hoist_data_discarding_ops.h:
```diff
@@ -0,0 +1,52 @@
+
+// This optimization hoists the data discarding ops (such as `skip`, `take` and
+//  `shard`) to avoid unnecessary computation.
+class HoistDataDiscardingOps : public TFDataOptimizerBase {
```
renamed.

zhuzilin committed:
updates based on reviews

zhuzilin commented on tensorflow/core/grappler/optimizers/data/hoist_data_discarding_ops.cc:
```diff
@@ -0,0 +1,120 @@
+
+constexpr std::array<const char*, 6> kCardinalityPreserving = {
+    "CacheDataset", "CacheDatasetV2", "PrefetchDataset",
+    "MapDataset", "ParallelMapDataset", "ParallelMapDatasetV2",
```
fixed.

zhuzilin commented on tensorflow/core/grappler/optimizers/data/hoist_data_discarding_ops.cc:
```diff
@@ -0,0 +1,120 @@
+    }
+  }
+  return false;
+}
```
changed.

zhuzilin committed:
update misleading doc

zhuzilin committed:
add python test

zhuzilin commented:
@jsimsa 
Thank you for your quick review! I've added the python test and updated the code according to the review. Could you have a second look?

aaudiber commented on tensorflow/python/data/experimental/kernel_tests/optimization/hoist_discard_test.py:
```diff
@@ -0,0 +1,71 @@
+from tensorflow.python.platform import test
+
+
+class HoistDiscardTest(test_base.DatasetTestBase, parameterized.TestCase):
```
Add tests for hoisting before prefetch, and for not hoisting before non-cardinality-preserving ops

aaudiber commented on tensorflow/python/data/experimental/ops/optimization_options.py:
```diff
@@ -126,6 +126,14 @@ class OptimizationOptions(options.OptionsBase):
+      ty=bool,
+      docstring=
+      "Whether to hoist ops that will discard data (such as skip, take, shard) "
+      "out of unary cardinality preserved transformations. "
```
Add a basic example, e.g. "dataset.map(...).take(3) gets optimized to dataset.take(3).map(...)".

aaudiber reviewed (commented):
Looks great overall, thanks for contributing this @zhuzilin!

zhuzilin committed:
modification based on reviews

zhuzilin commented on tensorflow/core/grappler/optimizers/data/hoist_discard.cc:
```diff
@@ -0,0 +1,116 @@
+namespace grappler {
+namespace {
+
+const std::unordered_set<string> kDataDiscarding = {
```
Changed.

zhuzilin commented:
@aaudiber 
Thank you for your reviews. I've changed the code according to them. Could you have a second look?

karmel commented:
For @tensorflow/api-owners :

We have a number of questions about this new option; can we make sure that the docstrings address:

1. What exactly is hoisting in this context? What are things hoisted into, and how does this affect the data pipeline in question? (Looking at the one-line example, this seems to be a reordering.)
2. What is the full set of transformations that is hoisted? Do these eventually get run somewhere else?
3. Can you add examples to the docstrings so that users will understand how to use this?

More broadly, should this be turned on by default? What would we have to test/ensure first?

zhuzilin commented:
@karmel 
Thank you for your questions :).
> 1. What exactly is hoisting in this context? What are things hoisted into, and how does this affect the data pipeline in question? (Looking at the one-line example, this seems to be a reordering.)

I'm a little confused about the terms reordering and hoisting. Should hoisting have loop involve? In that case, we may rename it to reordering. I was just imitating the pass `hoist_random_uniform`.

> 2. What is the full set of transformations that is hoisted? Do these eventually get run somewhere else?

The transformation should involve moving `skip`, `take`, `shard` to the front of `map`, `cache`, `prefetch` so that no unnecessary calculation or cache will take place. I think it will only involve tf.data ops.

> 3. Can you add examples to the docstrings so that users will understand how to use this?

I'd love to. Could you tell me how specific the docstrings should be? And should the extra docstrings be added in `optimization_options.py`?

> More broadly, should this be turned on by default? What would we have to test/ensure first?

I believe there will be no harm turning it on by default. And I'll be glad if it does.

@aaudiber @jsimsa Could you share your opinions on these questions? Thank you.

aaudiber commented:
> 1. What exactly is hoisting in this context? What are things hoisted into, and how does this affect the data pipeline in question? (Looking at the one-line example, this seems to be a reordering.)

I agree that reordering is a better term. Hoisting usually means moving to a different scope, as happens with `hoist_random_uniform`. How about calling it `reorder_data_discarding_ops`?

> 2. What is the full set of transformations that is hoisted? Do these eventually get run somewhere else?

@zhuzilin please update the docstring in optimization_options.py to specify the exact list of which transformations will be reordered, and also to make it clear that the optimization is only for performance, and will not affect the output of the dataset.

> 3. Can you add examples to the docstrings so that users will understand how to use this?
> 
> More broadly, should this be turned on by default? What would we have to test/ensure first?

It makes sense to turn this on by default, since it is almost always a strict improvement. I see two risks: (1) The user relies on a side-effect of applying their map function to discarded elements and (2) the user expects their entire dataset to be cached to disk, but with the re-ordering only part of the dataset is cached.

(1) is unlikely, and the user would be relying on undefined behavior.
(2) is more problematic - I think we should avoid applying the reordering to `cache` transformations which cache to a file instead of in-memory.

Once we've addressed the file-caching issue, lets turn the optimization by default, and I will run extra internal tests to make sure nothing is broken.

zhuzilin committed:
rename to reorder_data_discarding_ops

zhuzilin committed:
add details to doc

zhuzilin commented:
@aaudiber 
I've renamed the pass to `reorder_data_discarding_ops` and modified the doc. Could you have another look? As for the file-caching issue, I think we should maintain the optimization to it and tell user the consequence of it. Because in my opinion, using file cache to save the whole dataset while discarding some of it is not likely a common usage.
@karmel 
Could you also check the updated doc to see if it has the information we need?

aaudiber commented:
@zhuzilin Regarding cache-to-file, I think a common use case while debugging could be to add a `take()` after the call to `cache(filename)` to inspect some example data. For example, the user builds a dataset with

```
ds = create_dataset()
ds = ds.cache(filename)
```

Then they append `.take(3)` to make sure their data looks correct:

```
ds = create_dataset()
ds = ds.cache(filename)
for element in ds.take(3):
  print(element)
```

Then they remove the printing and run their training:

```
ds = create_dataset()
ds = ds.cache(filename)
train_on_dataset(ds)
```

The user will be surprised and unhappy that their dataset now produces only 3 elements.

zhuzilin commented:
@aaudiber 
Thank you for the information! I haven't thought about this use case before. However, it's hard to distinguish file cache and memory cache because `filename` is an input instead of a attribute, which means we can't known whether it will be empty or not in grappler... If we need to preserve the use case you mentioned, maybe we have to remain this optimization as default false?

aaudiber commented:
@zhuzilin 
I think it's important for the optimization to be on by default, so that it can have more impact. If we leave it off by default, not many users will be aware of it, and there will be very little usage. What do you think of removing the interaction with `cache` for now, and only re-ordering `map` and `prefetch`? Then later on we can split `CacheDataset` into separate `MemoryCacheDataset` and `FileCacheDataset` ops, so that we can include `MemoryCacheDataset` in this optimization.

zhuzilin commented:
@aaudiber 
Sure. I'll remove the `cache` for now if we will separate memory and file cache ops.

zhuzilin committed:
remove support for cache op

zhuzilin commented:
@aaudiber 
I've removed `cache` and updated the docstring. Could you have another look? :).

aaudiber commented on tensorflow/python/data/experimental/ops/optimization_options.py:
```diff
@@ -188,11 +188,9 @@ class OptimizationOptions(options.OptionsBase):
-      "data will no longer be saved after this optimization."
-      "If None, defaults to False.")
+      "`map` and `prefetch`. And notice this optimization is only for"
+      "performance, it will not affect the output of the dataset."
```
each line needs a space at the end so that the last word of the line doesn't get combined with the first word of the next line.

aaudiber replied:
semicolon after performance instead of comma

aaudiber commented on tensorflow/python/data/experimental/ops/optimization_options.py:
```diff
@@ -188,11 +188,9 @@ class OptimizationOptions(options.OptionsBase):
       "cardinality preserved transformations, e.g. dataset.map(...).take(3)"
```
preserved -> preserving

aaudiber commented on tensorflow/python/data/experimental/ops/optimization_options.py:
```diff
@@ -188,11 +188,9 @@ class OptimizationOptions(options.OptionsBase):
-      "However, it will influence the cache to the file, for the unused"
-      "data will no longer be saved after this optimization."
-      "If None, defaults to False.")
+      "`map` and `prefetch`. And notice this optimization is only for"
```
Remove "And notice"

aaudiber reviewed (changes requested):
Thanks @zhuzilin! I left a couple last comments on the documentation string, otherwise looks good to me

zhuzilin committed:
modify docstring

zhuzilin committed:
fix lint error

zhuzilin commented:
@aaudiber 
There was a lint error (one line was longer than 80 characters after renaming to reorder_data_discarding_ops). I've fix it and could you help aprove this pr again? Thank you!

zhuzilin commented:
@aaudiber 
It is strange that Linux CPU test has failed... All I did last commit was adding a newline in python code to make it shorter than 80 characters, but the Linux CPU test turns from passed to failed...
The error is about vectorizer registry:
```
tensorflow/core/grappler/optimizers/data/vectorization/vectorizer_registry_test.cc:46
Expected equality of these values:
  ::tensorflow::Status::OK()
    Which is: OK
  (s)
    Which is: Not found: Op type not registered '' in binary running on localhost.
  Make sure the Op and Kernel are registered in the binary running in this process.
  Note that if you are loading a saved graph which used ops from tf.contrib, accessing
  (e.g.) `tf.contrib.resampler` should be done before importing the graph,
  as contrib ops are lazily registered when the module is first accessed.
```
Is there any reason this new optimization may influence it?

aaudiber commented:
@zhuzilin That test is currently broken, unrelated to your PR. There is an outstanding change to disable the test while it is fixed.

karmel reviewed (approved):
for @tensorflow/api-owners

zhuzilin commented:
Genlty ping @aaudiber. I wonder if we are going to merge this pass? Thank you :).

zhuzilin commented:
@aaudiber 
The building failure is at `tensorflow/python/kernel_tests:cast_op_test`, which should have nothing to do with tf.data. Is there anything else I can help with this pr?

zhuzilin commented:
@aaudiber 
It seems that random building errors keep on happening 😂.

aaudiber reviewed (approved):
Trying again...

zhuzilin commented:
@aaudiber The tests are finally passed! 😂  Thank you so much for your help in this pr!

zhuzilin commented:
@gbaned @aaudiber Could you help to merge this pr? Thank you!

aaudiber commented:
@zhuzilin This should merge soon, I just need to fix up a couple internal tests
